### DISCUSSION TITLE
Anime4K for 3D Games

### DISCUSSION THREAD

**Discussion by bloc97** [Blinue/Magpie#183] [General]
First of all, thanks for creating Magpie, it's great for upscaling visual novels, but I think there's a lack of methods to upscale 3D games.

To help improve this software, I created two very optimized shaders for magpie based on the Anime4K CNN architecture. These two shaders are even smaller (about half the size, ~500 parameters) compared to `S` variants of Anime4K.
One is with anti-aliasing (for older games without AA), the other is for when the game already has anti-aliasing.
[Anime4K_3D.zip](https://github.com/Blinue/Magpie/files/7438838/Anime4K_3D.zip)

Upscaling from 1080p to 4K takes about ~0.9ms (1100 fps) on my Vega64 GPU, so I hope this will run well on low end systems/laptops.

One downside is the shaders perform poorly on the UI of games (a bit of ringing), this can't be fixed unless there's a way to upscale the UI separately.

Game without AA (Left = Original, Right = Upscaled x2):
![image](https://user-images.githubusercontent.com/567732/139375637-967f5a94-0d30-45d1-8fee-8b97333be8c4.png)![image](https://user-images.githubusercontent.com/567732/139375650-684fed22-4eb8-4491-a0d8-8f2c1524c979.png)

Game with AA:
![image](https://user-images.githubusercontent.com/567732/139375979-1f704d32-82f6-4feb-bdea-4eaac2656c77.png)![image](https://user-images.githubusercontent.com/567732/139375987-a5027d89-4369-4741-a5bc-99ac7ee05ae1.png)

If you were to compare this against DLSS v2.0, the quality is not great, but in my opinion the upscaling effect is better than bicubic or integer.

**Comment by Blinue** [comment]
Thanks for the nice work! You should try FSR, which is an alternative to DLSS, but much faster. It also supports scaling at any ratio. Many people use Magpie to improve the performance of 3D games.

[Comparison](https://imgsli.com/NzkyMzc/0/1)

**Reply by bloc97** [reply]
Yeah, after some testing I discovered that the non-AA model looks very much like FSR, but without noise amplification and with slight amounts of ringing.
I initially only trained the shader with AA (FSR cannot handle aliased games), but also trained the non-AA shader to see how it would perform.

**Reply by Blinue** [reply]
FSR performs scaling using traditional methods and therefore has superior performance that machine learning cannot match. However, the advantage of ML is that it can produce better results, which would be the most important reason for people to use it.

Hopefully, your work will provide a new option for 3D gamers :^)
